### PR TITLE
feat: detect thread starvation on internal dispatcher

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
@leviramsey this includes your previous contribution

File: akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala [hunk 1]
```diff
@@ -144,19 +148,46 @@ object StarvationDetector {
...
+
   /**
    * Creates and runs a StarvationDetector thread for the given ExecutionContext. Thread analytics are currently only
    * available for Akka dispatchers.
    *
    * You need to provide a `hasTerminated` function that will be used to figure out if the execution context has shut
    * down to shutdown the starvation detector thread.
+   *
+   * Note that an ExecutionContext will only have one StarvationDetector for it active at a time. If there is another
+   * StarvationDetector running, this method does nothing.
```
Comment (leviramsey) [inline akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala:183]: Having at times gone crazy adding StarvationDetectors to already-monitored contexts, this is welcome!

File: akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala [hunk 2]
```diff
@@ -144,19 +148,46 @@ object StarvationDetector {
...
    */
   def checkExecutionContext(
       ec: ExecutionContext,
       log: LoggingAdapter,
       config: StarvationDetectorSettings,
       hasTerminated: () => Boolean): Unit =
-    if (config.isEnabled) {
+    if (config.isEnabled && !starvationMonitoredContexts.contains(ec)) {
       val thread = new StarvationDetectorThread(ec, log, config, hasTerminated)
```
[the remark this answers is not in the document]
Comment (leviramsey) [reply akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala:191]: Could guard against with a `ConcurrentHashMap[ExecutionContext, StarvationDetectorThread]` and `putIfAbsent` (with the `StarvationDetectorThread` stopping itself if there's already a thread)

**Review by johanandren — APPROVED** [review]

File: akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-diagnostics/src/main/scala/akka/diagnostics/StarvationDetector.scala]: I agree, probably low risk but let's make it atomic: https://github.com/akka/akka-diagnostics/pull/10/commits/9b5a34dd41e37fb3a67042f620f4deeb964b578d

**Comment by patriknw** [comment]
unrelated flaky tests, I'll look at those separately

**Pull request merged**